### ISSUE TITLE
scipy.cluster.hierarchy.linkage fails silently when supplied a matrix with NaN row and column

### ISSUE THREAD

**Issue by chiffa** [scipy/scipy#5142]
When a matrix is supplied to `scipy.cluster.hierarchy.linkage(Distance_matrix, method="centroid")`, it will mostly fail without any error message, but terminate the Python process with a cryptic exit message, eg. `Process finished with exit code -1073740940 (0xC0000374)`.

In some cases `scipy.cluster.hierarchy.linkage` will succeed and when obtained linkage matrix is supplied to `scipy.cluster.hierarchy.fcluster(Linkage_Matrix, 0.95, criterion='distance')`, `fcluster` will fail with several possible errors: `ValueError: Linkage 'Z' contains negative counts`., `Linkage 'Z' uses the same cluster twice`.

**Comment by WarrenWeckesser** [comment]
Example:

```
In [3]: import scipy

In [4]: scipy.__version__
Out[4]: '0.17.0.dev0+845e465'
```

Create a distance matrix with a row and column of `nan`s:

```
In [5]: from scipy.spatial import distance_matrix

In [6]: np.random.seed(1234)

In [7]: x = np.random.rand(8, 2)

In [8]: d = distance_matrix(x, x)

In [9]: d[:, 2] = np.nan

In [10]: d[2, :] = np.nan
```

Now crash the python interpreter:

```
In [11]: from scipy.cluster.hierarchy import linkage

In [12]: linkage(d, method='centroid')
/Users/warren/anaconda/bin/python.app: line 3: 43680 Segmentation fault: 11  /Users/warren/anaconda/python.app/Contents/MacOS/python "$@"
```

**Comment by mp4096** [comment]
Same crash in scipy version `0.18.0`.

Create file `linkage_crash.py`:
```python
import numpy as np
from scipy.spatial import distance_matrix
from scipy.cluster.hierarchy import linkage

x = np.random.rand(8, 2)
d = distance_matrix(x, x)
d[:, 2] = np.NAN
d[2, :] = np.NAN
linkage(d, method='centroid')
```
and run it with `python -X faulthandler linkage_crash.py`. Segfault in [`hierarchy.py` at line 667](https://github.com/scipy/scipy/blob/maintenance/0.18.x/scipy/cluster/hierarchy.py#L667).

Now change the linkage method to `average`. Everything fine.

However, if I call `linkage` through `seaborn`, it segfaults for `method='average'`!
```python
import seaborn as sns
import numpy as np

a = np.empty((2, 2))
a[:] = np.NAN
sns.clustermap(a)
```
Segfault in [`hierarchy.py` at line 665](https://github.com/scipy/scipy/blob/maintenance/0.18.x/scipy/cluster/hierarchy.py#L665).

So the problem seems to be is somewhere in [`_hierarchy.pyx`](https://github.com/scipy/scipy/blob/maintenance/0.18.x/scipy/cluster/_hierarchy.pyx), in `linkage` as well as in `nn_chain`.

**Comment by nmayorov** [comment]
First I want to notice that `linkage` doesn't work with full distance matrix, only with 1-d condensed distance array or with raw vectors. So if you pass a 2-d array it will treat it like vectors and compute `pdist` from them.

Will be it sufficient to just check that all values in `y` are finite? Or does someone expect more sophisticated handling of NAN values?

**Comment by mp4096** [comment]
Oh, indeed, thanks for the hint.

So this works:
```python
import numpy as np
from scipy.spatial import distance_matrix
from scipy.cluster.hierarchy import linkage

x = np.random.rand(6)
x[0] = np.NAN
print(x.shape)
linkage(x, method='centroid')
```
and this segfaults:
```python
import numpy as np
from scipy.spatial import distance_matrix
from scipy.cluster.hierarchy import linkage

x = np.random.rand(6)
x[:] = np.NAN
print(x.shape)
linkage(x, method='centroid')
```

Apparently, all elements of the 1-d distance array must be `NaN` to cause a segfault. I didn't get a segfault if at least one element is not `NaN`.

**Comment by person142** [comment]
At least one issue is here:

https://github.com/scipy/scipy/blob/maintenance/0.18.x/scipy/cluster/_hierarchy.pyx#L836

i.e. in the block

``` python
while True:
    x = cluster_chain[chain_length - 1]
    current_min = NPY_INFINITYF

    for i in range(n):
        if size[i] == 0 or x == i:
            continue

        dist = D[condensed_index(n, x, i)]
        if dist < current_min:
            current_min = dist
            y = i

    if chain_length > 1 and y == cluster_chain[chain_length - 2]:
        break

    cluster_chain[chain_length] = y
    chain_length += 1
```
If `dist` is always `nan` then the conditional `if dist < current_min` is never true, so `y` never gets set. Then `if chain_length > 1 and y == cluster_chain[chain_length - 2]` is never true, leading to a segfault.

**Comment by mp4096** [comment]
Yes, that makes sense. An array filled with `np.inf`s segfaults as well.

@nmayorov: I'm not acquainted with clustering, but my intuitive suggestion would be to raise an exception if _all_ entries of `y` are either `np.nan` or positive `np.inf`. There could be some use cases in which `y` is filled with `np.nan`s, but even then an exception is better than a segfault.

**Comment by nmayorov** [comment]
A straightforward and reasonable solution is to allow only finite number. 

Say we want to handle NANs somehow. The distances between some points are not defined, how do we perform clustering then? I guess it doesn't make sense to pass distances with NANs to a clustering algorithm and just see what it returns (in case it doesn't crush). On the other hand, I can't think of any reasonable and simple way to handle NANs except removing each point which have a NAN distance to any other point. But it will break the numbering in dendrogram and most likely all algorithm in `scipy.cluster.hierarchy` will break. 

All in all, I suggest to require distances / vectors to contain only finite numbers and raise ValueError otherwise.

**Comment by mp4096** [comment]
Valid points! On the implementation side it would be fairly easy, I can make a pull request. 

On the other hand, I still have some reservations about this approach. 

First, the current behaviour is that `_hierarchy.pyx/nn_chain` segfaults only if a list with exactly one `np.inf` or `np.nan` element is given and `_hierarchy.pyx/linkage` segfaults only if a list full of `np.inf`s or `np.nan`s is given. In other cases, they just return NaNs and Infs. I am afraid this change can break a lot of existing code downstream. However, this definitely would force people to think more about their code and be more careful about data they're handling.

Second, the MATLAB implementation of `linkage` seems to have no problems with non-finite numbers and returning garbage:
```
>> linkage([Inf, Inf, Inf])
ans =
     2     3   Inf
     1     4   Inf
>> linkage([NaN, NaN, NaN])
ans =
     1     2   NaN
     3     4   NaN
```
I do not think that one should replicate MATLAB's behaviour, but I suppose they had their reasons to implement it like this.

**Comment by nmayorov** [comment]
Fixed in #6507.
